wicketd/src/installinator_progress.rs

```diff
@@ -0,0 +1,277 @@
+                }
+            }
+
+            ProgressReportStatus::Processed
+        } else {
+            slog::debug!(self.log, "update ID unrecognized"; "update_id" => %update_id);
+            ProgressReportStatus::UnrecognizedUpdateId
+        }
+    }
+}
+
+/// The update tracker's interface to the progress store.
+#[derive(Debug)]
+#[must_use]
+pub(crate) struct IprUpdateTracker {
+    log: slog::Logger,
+    running_updates: Arc<Mutex<HashMap<Uuid, RunningUpdate>>>,
+}
+
+impl IprUpdateTracker {
+    /// Registers an update ID and marks an update ID as started.
+    ///
+    /// Returns a oneshot receiver that resolves when the first message from the
+    /// installinator has been received.
+    pub(crate) async fn register(&self, update_id: Uuid) -> IprStartReceiver {
+        slog::debug!(self.log, "registering new update id"; "update_id" => %update_id);
+        let (start_sender, start_receiver) = oneshot::channel();
+
+        let mut running_updates = self.running_updates.lock().await;
+        running_updates.insert(update_id, RunningUpdate::Initial(start_sender));
+        start_receiver
+    }
+}
+
+/// Type alias for the receiver that resolves when the first message from the
+/// installinator has been received.
+pub(crate) type IprStartReceiver =
+    oneshot::Receiver<mpsc::Receiver<ProgressReport>>;
+
+#[derive(Debug)]
+#[must_use]
+enum RunningUpdate {
+    /// This is the initial state: the first message from the installinator
+    /// hasn't been received yet.
+    Initial(oneshot::Sender<mpsc::Receiver<ProgressReport>>),
+
+    /// Reports from the installinator have been received.
+    ReportsReceived(mpsc::Sender<ProgressReport>),
+
+    /// All messages have been received.
+    ///
+    /// We might receive multiple "completed" updates from the installinator in
+    /// case there's a network issue between wicketd and installinator. Build in
+    /// idempotency to ensure that the installinator doesn't fail for that
+    /// reason: rather than removing the UUID from self.running_updates once
+    /// it's done, we move to this state while keeping the UUID in the map.
+    Closed,
+
+    /// Temporary state used for on_report. Invalid outside on_report.
+    Invalid,
+}
+
+impl RunningUpdate {
+    fn take(&mut self) -> Self {
+        // Temporarily move to this state
+        std::mem::replace(self, Self::Invalid)
+    }
+
+    async fn send_and_next_state(
+        sender: mpsc::Sender<ProgressReport>,
+        report: ProgressReport,
+    ) -> Self {
+        let is_completed = Self::is_completed(&report);
+        _ = sender.send(report).await;
+        if is_completed {
+            Self::Closed
+        } else {
+            Self::ReportsReceived(sender)
+        }
+    }
+
+    fn is_completed(report: &ProgressReport) -> bool {
+        report
+            .completion_events
+            .iter()
+            // Scan in reverse order since the completion message is very likely
```

Comment:
Do we need to scan, or would it be sufficient to just check the last event?

Reply:
Done, I don't think there's any legitimate reason it wouldn't be the last event.